# Pre Entrega # 10
## Proyecto Data Science
##### Data Science - 32845
### José María Reyes Retana Tamayo
### Profesor: David Bustos Usta
### Tutor: Anahí García

## Resumen (Abstract):

El siguiente estudio se centra en el análisis de un conjunto de 3197 marcas de cervezas de 934 compañías productoras de cerveza que han sido evaluadas por usuarios a través de cuestionarios en línea, así mismo, se cuenta con la información reportada por las productoras sobre sus bebidas. Las cervezas son locales y son solamente de EEUU, del mismo modo, las evaluaciones son emitidas únicamente por consumidores de dicho país, por lo que este estudio no debe considerarse como válido para cualquier región fuera de los EEUU y más bien será tratado como una muestra pequeña dentro del universo cervecero pero que puede orientar a los lectores en cuanto a las tendencias y gustos vigentes.

La intención de este estudio ha sido la de comparar la información presentada por los productores de cerveza y comparar estos valores con las evaluaciones de los usuarios con la intención de encontrar que variables son las que más afectan a que una cerveza obtenga buenos resultados en el mercado de cervezas artesanales de los EEUU. Para lo anterior, se han realizado primero una indagación de los datos de forma gráfica a modo de análisis exploratorio de la información y obtención preliminar de resultados. Siguiente, se realizaron los cálculos matemáticos estadíticos para obtener valores de desviación estándar de las variables analizadas, coeficientes de correlación (donde así se propuso), prueba de hipótesis mediante el llamado "p valor", entre otros.

Las principales variables analizadas para este estudio han sido las relacionadas con el sabor de la cerveza: amargor y alcohol; y se han comparado con la evaluación que los consumidores han dado en las clasificaciones de: Aroma, Apariencia, Sensación en el Paladar, Sabor y Evaluación General.

A través de esta investigación, se ha observado que la calificación que un usuario da a cierta cerveza va estrechamente relacionada con la calificación que dará a la cerveza como conjunto; sin embargo, resulta difícil predecir cuál será la calificación de sabor de una cerveza.

## Índice:
1. Contexto del Estudio.\
    1.1 Contexto Comercial.\
    1.2 Contexto Analítico.
2. Conjunto de Datos.
3. Definición de las variables.
4. Análisis descriptivo de la información (EDA).
5. Hipótesis del proyecto.\
    5.1 Metodología de evaluación de las hipótesis.
6. Objetivo del proyecto.\
    6.1 Metodología propuesta.
7. Obtención de información.
8. Modelo de clasificación.\
    8.1 Selección de atributos (Feature Selection).\
    8.2 Creación y calibración del modelo de ML.\
    8.3 Prueba del modelo.\
    8.4 Métricas del modelo.
9. Modelo de regresión.\
    9.1 Selección de atributos (Feature Selection).\
    9.2 Creación y calibración del modelo de ML.\
    9.3 Prueba del modelo.\
    9.4 Métricas del modelo.
10. Conclusiones.\
    10.1 Conclusiones sobre las hipótesis.\
    10.2 Conclusiones sobre los modelos de ML.\
    10.3 Pensamientos finales.

## 1. Contexto del Estudio:

### 1.1 Contexto Comercial:

La industria de la cerveza es una de las pocas que pueden considerarse milenarias, el consumo de esta bebida está tan ligado a la humanidad como el pan mismo, siendo que prácticamente se dieron a la par. Sin embargo, si bien la bebida lleva ya milenios, literalmente, de consumo, sus estilos, presentaciones y hasta la forma y lugar donde se bebe han cambiado radicalmente, especialmente desde la invención de la refrigeración que ha permitido la exportación de estilos locales a todo el mundo y ha abierto el mercado de una forma sin precendente en la industria.

Por lo anterior, cada vez más, las distintas variedades de cerveza encuentran sus nichos de consumidores y se crean nuevos "sub-mercados" en los que la competencia ya no es contra todas las variedades, si no contra las de características similares por ser así de segmentados los gustos de los consumidores.

De acuerdo a lo ya mencionado, es importante que antes de comenzar con la producción con una nueva línea de cerveza, vale la pena comprender el nicho del mercado al que se quiere atender.

### 1.2 Contexto Analítico:

Se ha conseguido una base de datos que contiene más de tres mil marcas de cerveza diferentes, todas ellas de producción local (no en masa) en los Estados Unidos de América (EEUU).

La información no viene filtrada y se deberán realizar ajustes a las columnas y cálculos previos para poder mejorar la entendibilidad de los datos.

Se utilizarán procesos estadísticos para agrupar la información y para establecer patrones.

## Conjunto de Datos:

Se ha elegido trabajar con el data set nombrado "Beer Profiles and Ratings", obtenido de Kaggle en:
###### https://www.kaggle.com/code/aeryan/spotify-music-analysis/data

Creamos la base de datos a partir de un archivo CSV y la copiamos a un Data Frame (de Pandas) de trabajo, el cual modificaremos indiscriminadamente sin correr riesgo de alterar la información original.

In [26]:
# Importación de librerías iniciales:
import pandas as pd
import numpy as np

# Ubicación del archivo CSV dentro de el Codespace:
data_set_csv = "/workspaces/Coderhouse/Proyecto/Data/Raw/Beer Profiles and Ratings.csv"

# Creación del Data Frame:
data_set = pd.read_csv(data_set_csv)

print(type(data_set))
data_set.head()

<class 'pandas.core.frame.DataFrame'>


Name    Style  \
0                         Amber  Altbier   
1                    Double Bag  Altbier   
2                Long Trail Ale  Altbier   
3                  Doppelsticke  Altbier   
4  Sleigh'r Dark Doüble Alt Ale  Altbier   

                                            Brewery  \
0                               Alaskan Brewing Co.   
1                            Long Trail Brewing Co.   
2                            Long Trail Brewing Co.   
3  Uerige Obergärige Hausbrauerei GmbH / Zum Uerige   
4                           Ninkasi Brewing Company   

                                    Beer Name (Full)  \
0                  Alaskan Brewing Co. Alaskan Amber   
1                  Long Trail Brewing Co. Double Bag   
2              Long Trail Brewing Co. Long Trail Ale   
3  Uerige Obergärige Hausbrauerei GmbH / Zum Ueri...   
4  Ninkasi Brewing Company Sleigh'r Dark Doüble A...   

                                         Description  ABV  Min IBU  Max IBU  \
0  Notes:Richly malty and long on the palate, wit...  5.3       25       50   
1  Notes:This malty, full-bodied double alt is al...  7.2       25       50   
2  Notes:Long Trail Ale is a full-bodied amber al...  5.0       25       50   
3                                             Notes:  8.5       25       50   
4  Notes:Called 'Dark Double Alt' on the label.Se...  7.2       25       50   

   Astringency  Body  ...  Fruits  Hoppy  Spices  Malty  review_aroma  \
0           13    32  ...      33     57       8    111      3.498994   
1           12    57  ...      24     35      12     84      3.798337   
2           14    37  ...      10     54       4     62      3.409814   
3           13    55  ...      49     40      16    119      4.148098   
4           25    51  ...      11     51      20     95      3.625000   

   review_appearance  review_palate  review_taste  review_overall  \
0           3.636821       3.556338      3.643863        3.847082   
1           3.846154       3.904366      4.024948        4.034304   
2           3.667109       3.600796      3.631300        3.830239   
3           4.033967       4.150815      4.205163        4.005435   
4           3.973958       3.734375      3.765625        3.817708   

   number_of_reviews  
0                497  
1                481  
2                377  
3                368  
4                 96  

[5 rows x 25 columns]

Descripción del Data Frame donde se presenta la información con la que trabajaremos:

In [27]:
print(f"Forma del Data Frame: {data_set.shape}")
print(f"Número de dimensiones del Data Frame: {data_set.ndim}")
print(f"Cantidad de datos del Data Frame: {data_set.size}")
print(f"Nombre de las columnas del Data Frame: {data_set.columns}")

Forma del Data Frame: (3197, 25)
Número de dimensiones del Data Frame: 2
Cantidad de datos del Data Frame: 79925
Nombre de las columnas del Data Frame: Index(['Name', 'Style', 'Brewery', 'Beer Name (Full)', 'Description', 'ABV',
       'Min IBU', 'Max IBU', 'Astringency', 'Body', 'Alcohol', 'Bitter',
       'Sweet', 'Sour', 'Salty', 'Fruits', 'Hoppy', 'Spices', 'Malty',
       'review_aroma', 'review_appearance', 'review_palate', 'review_taste',
       'review_overall', 'number_of_reviews'],
      dtype='object')


In [28]:
# Se presentan el número de datos nulos en cada columna.
data_set.isnull().sum()

Name                 0
Style                0
Brewery              0
Beer Name (Full)     0
Description          0
ABV                  0
Min IBU              0
Max IBU              0
Astringency          0
Body                 0
Alcohol              0
Bitter               0
Sweet                0
Sour                 0
Salty                0
Fruits               0
Hoppy                0
Spices               0
Malty                0
review_aroma         0
review_appearance    0
review_palate        0
review_taste         0
review_overall       0
number_of_reviews    0
dtype: int64

Como podemos observar, no contamos con datos nulos en nuestra información y se tienen 3,197 entradas, cada una representando una cerveza diferente, misma que se ve descrita en 25 columnas en las que se especifican sus características.

In [29]:
# Método .info() para presentar la cantidad y el tipo de datos.
data_set.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3197 entries, 0 to 3196
Data columns (total 25 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               3197 non-null   object 
 1   Style              3197 non-null   object 
 2   Brewery            3197 non-null   object 
 3   Beer Name (Full)   3197 non-null   object 
 4   Description        3197 non-null   object 
 5   ABV                3197 non-null   float64
 6   Min IBU            3197 non-null   int64  
 7   Max IBU            3197 non-null   int64  
 8   Astringency        3197 non-null   int64  
 9   Body               3197 non-null   int64  
 10  Alcohol            3197 non-null   int64  
 11  Bitter             3197 non-null   int64  
 12  Sweet              3197 non-null   int64  
 13  Sour               3197 non-null   int64  
 14  Salty              3197 non-null   int64  
 15  Fruits             3197 non-null   int64  
 16  Hoppy              3197 

En la anterior tabla, podemos ver que la mayoría de nuestras variables son numéricas cuantitativas, lo cual nos permitirá trabajar más facilmente con modelos de Machine Learning sin necesidad de aplicar procesos de codificación de variables o etiquetado.

## 3. Definición de las variables:

Este data set contiene 25 columnas, las cuales se describen a continuación.

* Name: nombre de la cerveza, como se presenta en la etiqueta.

* Style: estilo de la cerveza.

* Brewery: nombre de la compañía cervecera que produce.

* Beer Name (Full): identificador único para cada cerveza, presenta nombre de la cerveza y la cervecera.

* Description: descripción de la cerveza de acuerdo al productor, no disponible para todas las cervezas presentadas.

* ABV: contenido de alcohol (% por volumen).
###### “International Bitterness Unit”, es la unidad utilizada para cuantificar que tan amarga es una cerveza. Los valores se presentan en un intervalo de 0 a 100, donde 0 es nada amargo y valores cercanos a 100 representan un alto grado de amargura.
* Min IBU: en esta columna se presentan los valores mínimos esperados de amargura.

* Max IBU: en esta columna se presentan los valores máximos esperados de amargura.
###### Las siguientes columnas presentan una calificación arbitraria otorgada por el autor del documento, que para este estudio se simplificará solo al valor numérico. Para cada uno de los aspectos evaluados por los consumidores. Entre mayor es la calificación, significa que dicha característica es más remarcable o evidente en la bebida.
* Astringency: astringencia.

* Body: cuerpo, que tan “densa” se siente en la boca.

* Alcohol: sensación de alcohol. No confundir con los grados de alcohol que es una medida objetiva, esta es la sensación que los consumidores tuvieron.

* Bitter: amargor. No confundir con los grados IBU que es una medida objetiva, esta es la sensación que los consumidores tuvieron.

* Sweet: dulzor.

* Sour: sabor ácido.

* Fruits: sabor afrutado.

* Hoppy: sabor a lúpulo.

* Spices: sabor a especias.

* Malty: sabor a malta (de cebada, trigo o cualquiera que sea el cereal con el que se elaboró la cerveza).
###### Las últimas seis columnas presentan la calificación general que los consumidores otorgaron a los distintos aspectos de la cerveza y el promedio general. Las calificaciones van del 0 al 5, puesto que la información fue obtenida de un portal donde se evaluaba mediante estrellas, siendo 5 el máximo valor posible.
* review _ aroma: evaluación del aroma (percepción olfativa) de la cerveza.

* review _ appearance: evaluación de la apariencia (percepción visual) de la cerveza.

* review _ palate: evaluación de la sensación en el paladar.

* review _ taste: evaluación del sabor (percepción del gusto) de la cerveza.

* review _ overall: evaluación global de la cerveza.
* number _ of _ reviews: total de evaluaciones recibidas.

## Análisis descriptivo de la información (EDA):

In [30]:
# Importamos las librerías a utilizar:
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

# Características para nuestros gráficos.
mpl.rcParams['axes.titleweight'] = 'heavy'

# Creamos el data frame de trabajo para evitar perder información:
wrk_df = data_set.copy()

print(type(wrk_df))

<class 'pandas.core.frame.DataFrame'>


In [31]:
for col in wrk_df.columns:
    if wrk_df[col].dtype == 'object':
        valores_unicos = wrk_df[col].nunique()
        print(f"Columna {col} tiene {valores_unicos} valores únicos.")

Columna Name tiene 3066 valores únicos.
Columna Style tiene 111 valores únicos.
Columna Brewery tiene 934 valores únicos.
Columna Beer Name (Full) tiene 3197 valores únicos.
Columna Description tiene 1841 valores únicos.


Puesto que ya hemos dado un vistazo a las distintas columnas y la forma en que los datos vienen presentados en el data frame original, se proponen algunos cambios a las columnas para poder procesar mejor la información.

Los cambios a realizar serán:
* Renombrar la columna Beer Name (Full) para hacerla más manejable.\
* Crear una columna con el valor promedio de la graduación IBU de las cervezas.\
    ** Se reporta un valor máximo y un valor mínimo de grados IBU en esta base de datos puesto que se trata de cervezas artesanales que aún son suceptibles a diferencias entre lote y lote.
* En base a investigación realizada para la mejor comprensión de los datos, se encuentra que un estándar en la industria cervecera es que una graduación de más de 35 grados IBU se considera una cerveza muy amarga y menor graduación que eso se considera poco amarga.\
* En base a investigación realizada para la mejor comprensión de los datos, se encuentra que un estándar en la industria cervecera es que una graduación de más de 6 grados de alcohol ABV se considera Alto y menos es lo normal.\
* Se separó la columna Style en dos, el principal y el secundario, ya que haciendo una revisión de esa columna se tenían demasiados estilos muy segmentados que dificultarían el modelo de clasificación.

In [32]:
# Renombramos la columna "Beer Name (Full)" para evitar errores de código:
wrk_df.rename(columns = {'Beer Name (Full)' : 'Full_name'}, inplace = True)

# Obtenemos un array con el promedio de IBU de cada cerveza.
prom_IBU = pd.DataFrame((wrk_df[["Min IBU", "Max IBU"]].mean(axis = 1)), columns = ['IBU_Promedio'])

# Agregamos la columna IBU Promedio a nuestra copia del Data Set.
wrk_df = pd.concat([wrk_df,prom_IBU], axis = 1)

# Creamos una columna que identique las cervezas si tienen más o menos de 35 Grados IBU:
wrk_df['amargor'] = np.where(wrk_df['IBU_Promedio'] > 35, 'Muy amarga', 'Poco Amarga')

# Separando la columna Style
wrk_df[["Style", "Secondary Style"]] = wrk_df.Style.str.split(" - ", expand = True)

# Creamos una columna que identique las cervezas si tienen más o menos de 6° de alcohol:
wrk_df['nivel_alcohol'] = np.where(wrk_df['ABV'] >= 6, 'Alto', 'Normal')

In [33]:
for col in wrk_df.columns:
    if wrk_df[col].dtype == 'object':
        valores_unicos = wrk_df[col].nunique()
        print(f"Columna {col} tiene {valores_unicos} valores únicos.")

Columna Name tiene 3066 valores únicos.
Columna Style tiene 43 valores únicos.
Columna Brewery tiene 934 valores únicos.
Columna Full_name tiene 3197 valores únicos.
Columna Description tiene 1841 valores únicos.
Columna amargor tiene 2 valores únicos.
Columna Secondary Style tiene 63 valores únicos.
Columna nivel_alcohol tiene 2 valores únicos.


Vemos como pasamos de 111 estilos distintos a solo 43, así como la clasificación en dos categorías para el IBU y el nivel de alcohol.